Mathlib/Combinatorics/Quiver/Path/Vertices.lean

```diff
@@ -160,7 +160,201 @@ lemma comp_eq_nil_iff {a b : V} {p : Path a b} {q : Path b a} :
+theorem exists_comp_of_mem_vertices {a b v : V} (p : Path a b)
+  (h : v ∈ p.vertices) : ∃ (p₁ : Path a v) (p₂ : Path v b), p = p₁.comp p₂ := by
+  obtain ⟨l₁, l₂, hv⟩ := (mem_iff_append.mp h : ∃ s t : List V, p.vertices = s ++ v :: t)
+  have h_len : l₁.length ≤ p.length := by
+    have : p.vertices.length = p.length + 1 := vertices_length p
+    have : l₁.length < p.vertices.length := by
+      rw [hv, length_append]
+      simp
+    omega
+  obtain ⟨c, p₁, p₂, hp, hl⟩ := exists_eq_comp_of_le_length p h_len
+  suffices hvc : v = c by
+    subst hvc
+    exact ⟨p₁, p₂, hp⟩
+  have h_verts : p.vertices = p₁.vertices.dropLast ++ p₂.vertices := by
+    rw [hp, vertices_comp]
+  have h_l1_len : l₁.length = p₁.vertices.dropLast.length := by aesop
+  have h_l1_eq : l₁ = p₁.vertices.dropLast := by
+    have : l₁ ++ v :: l₂ = p₁.vertices.dropLast ++ p₂.vertices := by
+      rw [← hv, h_verts]
+    exact append_inj_left this h_l1_len
+  have h_v_l2 : v :: l₂ = p₂.vertices := by
+    have : l₁ ++ v :: l₂ = p₁.vertices.dropLast ++ p₂.vertices := by
+      rw [← hv, h_verts]
+    rw [h_l1_eq] at this
+    exact append_cancel_left this
+  have : p₂.vertices.head? = some c := by
+    cases p₂ with
+    | nil => simp only [vertices_nil, head?_cons]
+    | cons _ _ => exact vertices_head? _
+  rw [← h_v_l2] at this
+  simp [head?_cons, Option.some.injEq] at this
+  exact this
+
+/-- `split_at_vertex` decomposes a path `p` at the vertex sitting in
+    position `i` of its `vertices` -/
+theorem split_at_vertices_index (i : ℕ)
+    (hi : i < p.vertices.length) :
+    ∃ (v : V) (p₁ : Path a v) (p₂ : Path v b),
+      p = p₁.comp p₂ ∧
+      p₁.length = i ∧
+      v = p.vertices[i] := by
+  have hi_le_len : i ≤ p.length := by
+    rw [vertices_length] at hi
+    exact Nat.le_of_lt_succ hi
+  obtain ⟨v, p₁, p₂, rfl, rfl⟩ := exists_eq_comp_of_le_length p hi_le_len
+  exact ⟨v, p₁, p₂, rfl, rfl, by simp⟩
+
+/-- Given vertices lists from a path composition,
+the prefix path’s vertices is a prefix of the full path’s vertices. -/
+lemma vertices_dropLast_isPrefix_of_vertices_eq
+    {V : Type*} [Quiver V]
+    {a b c : V} {p : Path a b} {p₁ : Path a c} {p₂ : Path c b}
+    (h : p.vertices = p₁.vertices.dropLast ++ p₂.vertices) :
+    p₁.vertices.dropLast.IsPrefix p.vertices := by
+  rw [h]; exact prefix_append p₁.vertices.dropLast p₂.vertices
+
+lemma mem_vertices_eq_end_or_mem_dropLast {a b c : V} (p : Path a b) (h : c ∈ p.vertices) :
+  c = b ∨ c ∈ p.vertices.dropLast := by
+  induction p with
+  | nil =>
+    simp [vertices_nil] at h
+    subst h
+    simp
+  | cons p' e ih =>
+    rename_i b'
+    have h' : c ∈ p'.vertices ∨ c = b' := by
+      simpa using (mem_vertices_cons p' e).1 h
+    cases h' with
+    | inl h_in_p' =>
+      have ih' := ih h_in_p'
+      cases ih' with
+      | inl h_eq_p'_end =>
+        right
+        subst h_eq_p'_end
+        simp
+      | inr h_in_p'_dropLast =>
+        simp_all
+    | inr h_eq_b =>
+      left
+      exact h_eq_b
```

Comment:
This is really a lemma about lists. There is just one detail which is special to the situation of paths in a quiver which is:
```lean
@[simp] lemma dropLast_append_end_eq :
    p.vertices.dropLast ++ [b] = p.vertices := by
  simp_rw [← p.vertices_getLast p.vertices_ne_nil, dropLast_concat_getLast]
```
Given this we can prove `mem_vertices_eq_end_or_mem_dropLast` quite efficiently as follows:
```lean
example {c : V} (h : c ∈ p.vertices) :
    c ∈ p.vertices.dropLast ∨ c = b := by
  rw [← dropLast_append_end_eq] at h
  simpa [-dropLast_append_end_eq] using h
```
So I suggest adding `dropLast_append_end_eq` (somewhere higher up in the file) and then dropping `mem_vertices_eq_end_or_mem_dropLast` altogether.